netctx/packetconnctx.go

```diff
@@ -0,0 +1,163 @@
+// SPDX-FileCopyrightText: 2023 The Pion community <https://pion.ly>
+// SPDX-License-Identifier: MIT
+
+package netctx
+
+import (
+	"context"
+	"io"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// ReaderFrom is an interface for context controlled packet reader.
+type ReaderFrom interface {
+	ReadFromContext(context.Context, []byte) (int, net.Addr, error)
+}
+
+// WriterTo is an interface for context controlled packet writer.
+type WriterTo interface {
+	WriteToContext(context.Context, []byte, net.Addr) (int, error)
+}
+
+// ReadFromWriterTo is a composite of ReadFromWriterTo.
+type ReadFromWriterTo interface {
+	ReaderFrom
+	WriterTo
+}
+
+// PacketConnCtx is a wrapper of net.PacketConn using context.Context.
+type PacketConnCtx interface {
+	ReaderFrom
+	WriterTo
+	io.Closer
+	LocalAddr() net.Addr
+	Conn() net.PacketConn
+}
+
+type packetConnCtx struct {
+	nextConn  net.PacketConn
+	closed    chan struct{}
+	closeOnce sync.Once
+	readMu    sync.Mutex
+	writeMu   sync.Mutex
+}
+
+// NewPacketConn creates a new PacketConnCtx wrapping the given net.PacketConn.
+func NewPacketConn(pconn net.PacketConn) PacketConnCtx {
+	p := &packetConnCtx{
+		nextConn: pconn,
+		closed:   make(chan struct{}),
+	}
+	return p
+}
+
+func (p *packetConnCtx) ReadFromContext(ctx context.Context, b []byte) (int, net.Addr, error) {
+	p.readMu.Lock()
+	defer p.readMu.Unlock()
+
+	select {
+	case <-p.closed:
+		return 0, nil, io.EOF
```

Comment:
Isn't a `net.ErrClosed` more appropriate here?

Reply:
That seems more reasonable to me 👍🏻